src/HexagonAlignLoads.cpp

```diff
@@ -0,0 +1,242 @@
+        //       modulus_remainder.remainder lanes off, but we are only 1 lane off.
+        int lanes = ramp->lanes;
+        Expr base = ramp->base;
+        // We know the base itself puts us off an aligned address by base_lanes_off
+        // number of lanes.
+        int base_lanes_off = mod_imp(host_alignment, lanes);
+        ModulusRemainder mod_rem = get_alignment_info(base);
+        if (mod_rem.modulus == 1 && mod_rem.remainder == 0) {
+            // We can't reason about alignment.
+            return AlignCheck::NoResult;
+        }
+        else {
+            int base_mod = base_lanes_off + mod_imp(mod_rem.modulus, lanes);
+            int rem_mod = mod_imp(mod_rem.remainder, lanes);
+            if (!(base_mod + rem_mod)) return AlignCheck::Aligned;
+            else {
+                if (lanes_off) *lanes_off = base_mod + rem_mod;
+                return AlignCheck::Unaligned;
+            }
+        }
+    }
+    void visit(const Load *op) {
+        debug(4) << "HexagonAlignLoads: Working on " << (Expr) op << "..\n";
+        Expr index = mutate(op->index);
+        if (op->type.is_vector()) {
+            bool external = op->image.defined();
+            if (external) {
+                debug(4) << "HexagonAlignLoads: Not dealing with an external image\n";
+                debug(4) << (Expr) op << "\n";
+                expr = op;
+                return;
+            }
+            else {
+                const Ramp *ramp = index.as<Ramp>();
+                const IntImm *stride = ramp ? ramp->stride.as<IntImm>() : NULL;
+                // We will work only on natural vectors supported by the target.
+                if (ramp->lanes != natural_vector_lanes(op->type)) {
+                    expr = op;
+                    return;
+                }
+                if (ramp && stride && stride->value == 1) {
+                    int lanes = ramp->lanes;
+                    // If this is a parameter, the base_alignment should be host_alignment.
+                    // This cannot be an external image because we have already checked for
+                    // it. Otherwise, this is an internal buffers that is always aligned
+                    // to the natural vector width.
+                    int base_alignment = op->param.defined() ?
+                        op->param.host_alignment() : vector_size;
+                    int lanes_off;
+                    AlignCheck ac = get_alignment_check(ramp, base_alignment, &lanes_off);
+                    if (ac == AlignCheck::NoResult) {
+                        // We know nothing about alignment. Give up.
+                        // TODO: Fix this.
+                        debug(4) << "HexagonAlignLoads: Cannot reason about alignment.\n";
+                        debug(4) << "HexagonAlignLoads: Type: " << op->type << "\n";
+                        debug(4) << "HexagonAlignLoads: Index: " << index << "\n";
+                        expr = op;
+                        return;
```

Comment:
It looks like the control flow here would be simpler if the if below were an else-if, and the return statements were pulled out of the branches.